File: backend/src/main/kotlin/no/uib/echo/schema/Degree.kt
package no.uib.echo.schema

enum class Degree {
    DTEK,
    DSIK,
    DVIT,
    BINF,
    IMO,
    INF,
    PROG,
    ARMNINF,
    POST,
    DSC,

    // MISC, IKT and KOGNI should not be used,
    // they are only here for backwards compatibility.
    MISC,
    IKT,
    KOGNI
}

val bachelors: List<Degree> = listOf(Degree.DTEK, Degree.DSIK, Degree.DVIT, Degree.BINF, Degree.IMO, Degree.DSC)
val masters: List<Degree> = listOf(Degree.INF, Degree.PROG, Degree.DSC)

fun nullableStringToDegree(str: String?): Degree? {
    return when (str) {
        "DTEK" -> Degree.DTEK
        "DSIK" -> Degree.DSIK
        "DVIT" -> Degree.DVIT
        "BINF" -> Degree.BINF
        "IMO" -> Degree.IMO
        "INF" -> Degree.INF
        "PROG" -> Degree.PROG
        "ARMNINF" -> Degree.ARMNINF
        "POST" -> Degree.POST
        "DSC" -> Degree.DSC
        // MISC, IKT and KOGNI should not be used,
        // they are only here for backwards compatibility.
        "MISC" -> Degree.MISC
        "IKT" -> Degree.IKT
        "KOGNI" -> Degree.KOGNI
        else -> null
    }
}

fun nullableDegreeToString(deg: Degree?): String? {
    return when (deg) {
        Degree.DTEK -> "DTEK"
        Degree.DSIK -> "DSIK"
        Degree.DVIT -> "DVIT"
        Degree.BINF -> "BINF"
        Degree.IMO -> "IMO"
        Degree.INF -> "INF"
        Degree.PROG -> "PROG"
        Degree.ARMNINF -> "ARMNINF"
        Degree.POST -> "POST"
        Degree.DSC -> "DSC"
        // MISC, IKT and KOGNI should not be used,
        // they are only here for backwards compatibility.
        Degree.MISC -> "MISC"
        Degree.IKT -> "IKT"
        Degree.KOGNI -> "KOGNI"
        else -> null
    }
}

File: frontend/src/lib/utils/degree.ts
import type { Degree } from '@utils/schemas';

const allDegrees: Array<Degree> = [
    'DTEK',
    'DSIK',
    'DVIT',
    'BINF',
    'IMO',
    'IKT',
    'KOGNI',
    'INF',
    'PROG',
    'ARMNINF',
    'POST',
    'MISC',
    'DSC',
];

const feideDegreeGroupPrefix = 'fc:fs:fs:prg:uib.no:';

const degreeToFeideGroup = (degree: Degree): string | null => {
    switch (degree) {
        case 'DTEK': {
            return feideDegreeGroupPrefix + 'BAMN-DTEK';
        }
        case 'DSIK': {
            return feideDegreeGroupPrefix + 'BAMN-DSIK';
        }
        case 'DVIT': {
            return feideDegreeGroupPrefix + 'BAMN-DVIT';
        }
        case 'BINF': {
            return feideDegreeGroupPrefix + 'BAMN-BINF';
        }
        case 'IMO': {
            return feideDegreeGroupPrefix + 'BATF-IMØ';
        }
        case 'INF': {
            return feideDegreeGroupPrefix + 'MAMN-INF';
        }
        case 'PROG': {
            return feideDegreeGroupPrefix + 'MAMN-PROG';
        }
        case 'ARMNINF': {
            return feideDegreeGroupPrefix + 'ÅRMN-INF';
        }
        case 'DSC': {
            return feideDegreeGroupPrefix + '5MAMN-DSC';
        }
        case 'POST': {
            return feideDegreeGroupPrefix + 'POST';
        }
        default: {
            return null;
        }
    }
};

const allValidDegrees: Array<Degree> = allDegrees.filter((degree) => degreeToFeideGroup(degree) !== null);

const allValidFeideGroups: Array<string> = allValidDegrees.reduce((acc, degree) => {
    const feideGroup = degreeToFeideGroup(degree);
    if (feideGroup !== null) {
        acc.push(feideGroup);
    }
    return acc;
}, [] as Array<string>);

export { allDegrees, allValidDegrees, allValidFeideGroups, degreeToFeideGroup };
